src/app-layer-dnp3.c

```diff
@@ -0,0 +1,2616 @@
+        a1 = header->src;
+        a2 = header->dst;
+    }
+
+    if (blen == sizeof(DNP3Session)) {
+        DNP3Session *session = b;
+        b1 = session->master;
+        b2 = session->slave;
+    }
+    else if (blen == sizeof(DNP3LinkHeader)) {
+        DNP3LinkHeader *header = b;
+        b1 = header->src;
+        b2 = header->dst;
+    }
+
+    if (a1 == b1 && a2 == b2) {
+        return 1;
+    }
+    if (a1 == b2 && a2 == b1) {
+        return 1;
+    }
+
+    return 0;
+}
+
+/**
+ * \brief Reassemble the application layer by stripping the CRCs.
+ *
+ * \param input Input buffer starting at the transport header (which
+ *    will be removed from the output).
+ * \param input_len Length of the input buffer.
+ * \param output Pointer to output buffer (may be realloc'd).
+ * \param output_len Pointer to output length.
+ *
+ * \retval Pointer to output buffer if successful, otherwise NULL.
+ */
+static uint8_t *DNP3ReassembleApplicationLayer(uint8_t *input,
+    uint32_t input_len, uint8_t **output, uint32_t *output_len)
+{
+    int len = (input_len / (DNP3_BLOCK_SIZE + DNP3_CRC_LEN)) * DNP3_BLOCK_SIZE;
+    int rem = input_len % (DNP3_BLOCK_SIZE + DNP3_CRC_LEN);
+
+    if (rem) {
+        /* The remainder must be at least one byte plus a CRC. */
+        if (rem < 3) {
+            return NULL;
+        }
+        len += rem + DNP3_CRC_LEN;
+    }
+
+    /* Remove one byte for the transport header which won't be
+     * included in the output. */
+    len--;
+
+    if (*output == NULL) {
+        *output = SCCalloc(1, len);
+        if (unlikely(*output == NULL)) {
+            return NULL;
+        }
+    }
+    else {
+        uint8_t *ptr = SCRealloc(*output, (size_t)(*output_len + len));
+        if (unlikely(ptr == NULL)) {
+            SCFree(*output);
+            return NULL;
+        }
+        *output = ptr;
+    }
+
+    int offset = 0, block_size;
+    while ((uint32_t)offset < input_len) {
+        if (input_len - offset > DNP3_BLOCK_SIZE + DNP3_CRC_LEN) {
+            block_size = DNP3_BLOCK_SIZE + DNP3_CRC_LEN;
+        }
+        else {
+            block_size = input_len - offset;
+        }
+
+        /* For the first block we trim the leading transport header. */
+        if (offset == 0) {
+            memcpy(*output + *output_len, input + offset + 1,
+                block_size - DNP3_CRC_LEN - 1);
+            *output_len += block_size - DNP3_CRC_LEN - 1;
+        }
+        else {
+            memcpy(*output + *output_len, input + offset,
+                block_size - DNP3_CRC_LEN);
+            *output_len += block_size - DNP3_CRC_LEN;
+        }
+        offset += block_size;
+    }
+
+    return *output;
+}
+
+static void *DNP3StateAlloc(void)
+{
+    SCEnter();
+    DNP3State *dnp3;
+
+    dnp3 = (DNP3State *)SCCalloc(1, sizeof(DNP3State));
+    if (unlikely(dnp3 == NULL)) {
+        return NULL;
+    }
+    TAILQ_INIT(&dnp3->tx_list);
+
+    dnp3->sessions = HashListTableInit(256, DNP3SessionHash,
```

Comment:
This will lead to a lot of memory use. The struct itself is 48 bytes, but the 256 means we're doing 256 * 8 bytes of allocations. If the number of session won't be high this isn't a problem necessarily, but it's something to keep in mind.